### PR TITLE
SALTO-790: change adapter api to a single deploy endpoint

### PR DESCRIPTION
**Pull request by ori-moisis**

### CODE REVIEW

File: packages/salesforce-adapter/src/adapter.ts [hunk 1]
```diff
@@ -355,23 +356,36 @@ export default class SalesforceAdapter {
     return { elements, updatedConfig: { config, message: STOP_MANAGING_ITEMS_MSG } }
   }
 
+  async deploy(changes: ChangeGroup): Promise<DeployResult> {
```
Comment (omrilit) [inline packages/salesforce-adapter/src/adapter.ts:359]: nit: I think that it's more intuitive to name this arg `changeGroup` and then avoid the confusing `changes.changes` (I saw it in several places in the PR)

File: packages/adapter-utils/src/deploy.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+*/
+import { ChangeGroup, DeployResult, isInstanceChange, InstanceElement } from '@salto-io/adapter-api'
+
+export type ChangeOperations = {
+  add: (elem: InstanceElement) => Promise<InstanceElement>
+  remove: (elem: InstanceElement) => Promise<void>
+  update: (before: InstanceElement, after: InstanceElement) => Promise<InstanceElement>
+}
+
+export const deployInstance = async (
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/adapter-utils/src/deploy.ts:24]: I don't think this function should exist at all in the long run.
in my mind it is only here to help avoid refactoring the adapters in the same PR as all the other changes.
there is no reason to keep the old interface in adapters, it just makes the code more complicated than it has to be.

File: packages/netsuite-adapter/test/adapter.test.ts [hunk 3]
```diff
@@ -161,21 +178,35 @@ describe('Adapter', () => {
...
         const instWithUnsupportedType = new InstanceElement('unsupported',
           new ObjectType({ elemID: new ElemID(NETSUITE, 'UnsupportedType') }))
-        await expect(netsuiteAdapter.add(instWithUnsupportedType)).rejects.toThrow()
+        const result = await adapterAdd(instWithUnsupportedType)
+        expect(result.appliedChanges).toHaveLength(0)
+        expect(result.errors).toHaveLength(1)
       })
 
-      it('should throw error when trying to add a typesToSkip instance', async () => {
+      it('should fail when trying to add a customTypesToSkip instance', async () => {
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/netsuite-adapter/test/adapter.test.ts:189]: I think I just made a mistake in merge conflict resolution

File: packages/netsuite-adapter/test/adapter.test.ts [hunk 4]
```diff
@@ -121,8 +121,16 @@ describe('Adapter', () => {
       client.deployFolder = jest.fn().mockImplementation(() => Promise.resolve())
     })
     describe('add', () => {
+      const adapterAdd = (after: ChangeDataType): Promise<DeployResult> => netsuiteAdapter.deploy({
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/netsuite-adapter/test/adapter.test.ts:124]: plan to do it

File: packages/core/src/api.ts [hunk 5]
```diff
@@ -142,7 +164,10 @@ export const deploy = async (
       elementsAfterHiddenRemoval,
       workspaceElements
     )).map(change => ({ change, serviceChange: change }))
-      .map(toChangesWithPath(name => changedElementMap[name] || []))
+      .map(toChangesWithPath(name => {
+        const elem = changedElements.get(name)
+        return elem === undefined ? [] : [elem]
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/core/src/api.ts:169]: changed to:
```
.map(toChangesWithPath(name => collections.array.makeArray(changedElements.get(name))))
```
instead

File: packages/core/src/core/deploy.ts [hunk 6]
```diff
@@ -59,18 +56,18 @@ export type StepEvents = {
 
 export const deployActions = async (
   deployPlan: Plan,
-  adapters: Record<string, Adapter>,
+  adapters: Record<string, AdapterOperations>,
   reportProgress: (item: PlanItem, status: ItemStatus, details?: string) => void,
-  postApplyAction: (action: ActionName, element: Element) => Promise<void>
+  postApplyAction: (appliedChanges: ReadonlyArray<Change>) => Promise<void>
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/core/src/core/deploy.ts:61]: changed to `postDeployAction`

**Pull request merged**